R/functions.R

```diff
@@ -185,75 +205,564 @@ getAtlasData <- function( experimentAccessions ) {
+
+        # Create a named vector with id as names and label as values
+        assay_vector <- setNames(xml_attr(assay_groups, "label"), xml_attr(assay_groups, "id"))
+
+        # Replace the columns names in results, excluding 'GeneID' and 'Gene.Name' columns
+        colnames(results)[-c(1, 2)] <- assay_vector[colnames(results)[-c(1, 2)]]
+
+        return(results)
+
+    } else if (normalisation %in% c("cpm")) {
+        # Define filenames for TPM and FPKM files.
+
+        experiment_summary <- getAtlasData( c(experimentAccession))[[experimentAccession]]
+
+        # Get counts.
+        counts_matrix <- assay(experiment_summary$rnaseq, "counts")
+
+        # Calculates CPM from counts using EdgeR.
+        cpm_matrix <- edgeR::cpm(counts_matrix)
+
+        # Converts to DataFrame to match TPM and FPKM output.
+        cpm_df <- as.data.frame(cpm_matrix)
+
+        # Adds two columns GeneID and Gene.Name to DataFrame to match TPM and FPKM output format.
+        # cpm_df <- cbind(GeneID = rownames(cpm_df), Gene.Name = NA, cpm_df)
+
+        xmlContent <- .downloadXMLFile(xmlUrl)
+
+        assay_groups <- xml_find_all(xmlContent, ".//assay_group")
+
+        subgroups_list <- list()
+
+        # Loop through each assay group and extract label and assays
+        for (group in assay_groups) {
+            # Get the label of the assay group
+            label <- xml_attr(group, "label")
+
+            # Get all assay values in the group
+            assays <- xml_text(xml_find_all(group, "assay"))
+
+            # Add to the list
+            subgroups_list[[label]] <- assays
+        }
+
+        cpm_mean_df <- data.frame(row.names = rownames(cpm_df))
+
+        # Loop through subgroups and calculate the row-wise mean
+        for (group_name in names(subgroups_list)) {
+            group_columns <- subgroups_list[[group_name]]
+
+            # Ensure columns exist in 'results' before proceeding
+            if(all(group_columns %in% colnames(cpm_df))) {
+                # Calculate the row-wise mean for the subgroup columns and assign it to df
+                cpm_mean_df[[group_name]] <- rowMeans(cpm_df[, group_columns, drop = FALSE])
+            } else {
+                print(paste("Error: Columns not found for", group_name))
+            }
+        }
+
+        cpm_mean_df <- cbind(GeneID = rownames(cpm_mean_df), Gene.Name = NA, cpm_mean_df)
+
+        # assayData(ex[["A-AFFY-44"]])$exprs 
+        # this is microarray so not applied here
+
+        return(cpm_mean_df)
+
+    } else {
+        stop("Defined Normalisation type ", normalisation ," is not supported, must be a tpm, fpkm or cpm.")
+    }
+
+}
+
+
+heatmapAtlasExperiment <- function(df, 
+                                     filename = "heatmap",
+                                     heatmap_color = "Blues",
+                                     top_n = 100,
+                                     show_heatmap_title = TRUE ) {  
+
```

Comment:
Hi @anilthanki I think what we want here in first instance is to display the plot rather than save it to a pdf file. I'll try to edit the function

Reply:
It would be good to add a legend to the heatmap too